File: cpp/230.cpp
#include<utility>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
  pair<int, int> kth(TreeNode* root, int k){
    int lcnt = 0, rcnt = 0, lval, rval;
    if(root->left){
      auto p = kth(root->left, k);
      lcnt = p.first; lval = p.second;
    }
    if(root->right){
      auto p = kth(root->right, k-lcnt-1);
      rcnt = p.first; rval = p.second;
    }
    int total = lcnt + 1 + rcnt;
    if(k <= lcnt) return {total, lval};
    else if (k == lcnt+1) return {total, root->val};
    else return {total, rval};
  }
public:
  int kthSmallest(TreeNode* root, int k) {
    return kth(root, k).second;
  }
};

File: python/230.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def kth(self, root, k):
        if root.left:
            lcnt, lval = self.kth(root.left, k)
        else:
            lcnt, lval = 0, None
        if root.right:
            rcnt, rval = self.kth(root.right, k-lcnt-1)
        else:
            rcnt, rval = 0, None
        total = lcnt + rcnt + 1
        if k <= lcnt:
            return total, lval
        elif lcnt+1 == k:
            return total, root.val
        else:
            return total, rval

    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        return self.kth(root, k)[1]
